source/brailleDisplayDrivers/albatross/driver.py

```diff
@@ -0,0 +1,641 @@
+		"""Handles display button presses.
+		in single ctrl-key presses and ctrl-key combinations the first key is
+		resent as last one.
+		"""
+		if ord(data) in CONTROL_KEY_CODES or self._waitingCtrlPacket:
+			# at most MAX_COMBINATION_KEYS keys.
+			if self._waitingCtrlPacket:
+				data = self._partialCtrlPacket + data
+			if len(data) > 1 and data[0] == data[len(data) - 1]:
+				pass  # enclosed ctrl packet
+			else:
+				log.debug(f"Partial ctrl packet {data}")
+				try:
+					while len(data) <= MAX_COMBINATION_KEYS:
+						oneKey = self._readQueue.popleft()
+						log.debug(f"Read: dequeued key {oneKey}, {len(self._readQueue)} items left")
+						# Encloses ctrl packet
+						if ord(oneKey) == data[0]:
+							# No reason but easier debugging to add oneKey to data
+							data += oneKey
+							break
+						data += oneKey
+				except IndexError:
+					self._waitingCtrlPacket = True
+					self._partialCtrlPacket = data
+					log.debug(
+						f"Read: Ctrl key packet {data} dequeued partially, "
+						"_readQueue is empty", exc_info=True
+					)
+					return
+				if len(data) > MAX_COMBINATION_KEYS and data[len(data) - 1] != data[0]:
+					self._waitingCtrlPacket = False
+					self._partialCtrlPacket = None
+					log.debug(f"Not valid key combination, ignoring {data}")
+					return
+		log.debug(f"Keys for key press: {data}")
+		pressedKeys = set(data)
+		log.debug(f"Forwarding keys {pressedKeys}")
+		try:
+			inputCore.manager.executeGesture(
+				gestures.InputGestureKeys(pressedKeys, self.name)
+			)
+		# Attribute error which rarely occurs here is something strange.
+		except (inputCore.NoInputGestureAction, AttributeError):
+			log.debug("", exc_info=True)
+		if self._waitingCtrlPacket:
+			self._waitingCtrlPacket = False
+			self._partialCtrlPacket = None
+
+	def _keepConnected(self):
+		"""Keep display connected if nothing is sent for a while."""
+		if self._tryToConnect or not self.numCells:
+			return
+		if time.time() - self._writeTime >= KC_INTERVAL:
+			self._writeQueue.append(BOTH_BYTES)
+			self._somethingToWrite()
+
+	def _clearOldCells(self):
+		"""Whole display should be updated at next time."""
+		for i in range(len(self._oldCells)):
+			self._oldCells[i] = 0
+
+	def display(self, cells: List[int]):
+		"""Prepare cell content for display.
+		@param cells: List of cells content."
+		"""
+		# Keep _currentCells up to date for reconnection regardless
+		# of connection state of driver.
+		self._currentCells = cells.copy()
+		# No connection
+		if self._tryToConnect or not self.numCells:
+			return
+		writeBytes: List[bytes] = [START_BYTE, ]
+		# Only changed content is sent (cell index and data).
+		for i, cell in enumerate(cells):
+			if cell != self._oldCells[i]:
+				self._oldCells[i] = cell
+				# display indexing starts from 1
+				writeBytes.append((i + 1).to_bytes(1, 'big'))
+				# Bits have to be reversed.
+				# Source: https://stackoverflow.com/questions/12681945/reversing-bits-of-python-integer

```

Comment:
these source comments for code snippets aren't helpful. relevant information should be self contained within comments and code should hold up on its own merit.